pybamm/solvers/idaklu_jax.py

```diff
@@ -0,0 +1,956 @@
+                output_variables if output_variables else self.jax_output_variables
+            )
+        except AttributeError:
+            raise pybamm.SolverError("jaxify() must be called before jax_value()")
+        d = {}
+        for outvar in output_variables:
+            d[outvar] = jax.vmap(
+                self.get_var(f, outvar),
+                in_axes=(0, None),
+            )(t, inputs)
+        return d
+
+    def jax_grad(
+        self,
+        *,
+        f: Union[Callable, None] = None,
+        t: np.ndarray = None,
+        inputs: Union[dict, None] = None,
+        output_variables: Union[List[str], None] = None,
+    ):
+        """Helper function to compute the gradient of a jaxified expression
+
+        Returns a numeric (np.ndarray) object (not a JAX expression).
+        Parameters are inferred from the base object, but can be overridden.
+
+        Parameters
+        ----------
+        f : function
+            The jaxified expression
+        t : float | np.ndarray, optional
+            Time sample or vector of time samples
+        inputs : dict, optional
+            dictionary of input values
+        output_variables : list of str, optional
+            The variables to be returned. If None, the variables in the model are used.
+        """
+        try:
+            f = f if f else self.jaxify_f
+            t = t if t else self.jax_t_eval
+            inputs = inputs if inputs else self.jax_inputs
+            output_variables = (
+                output_variables if output_variables else self.jax_output_variables
+            )
+        except AttributeError:
+            raise pybamm.SolverError("jaxify() must be called before jax_grad()")
+        d = {}
+        for outvar in output_variables:
+            d[outvar] = jax.vmap(
+                jax.grad(
+                    self.get_var(f, outvar),
+                    argnums=1,
+                ),
+                in_axes=(0, None),
+            )(t, inputs)
+        return d
+
+    class _hashabledict(dict):
+        def __hash__(self):
+            return hash(tuple(sorted(self.items())))
+
+    @lru_cache(maxsize=1)
+    def _cached_solve(self, model, t_hashable, *args, **kwargs):
+        """Cache the last solve for reuse"""
+        return self.solver.solve(model, t_hashable, *args, **kwargs)
+
+    def _jaxify_solve(self, t, invar, *inputs_values):
+        """Solve the model using the IDAKLU solver
+
+        This method is called by the JAX primitive definition and caches the last solve
+        fo reuse.
+        """
+        # Reconstruct dictionary of inputs
+        if self.jax_inputs is None:
+            d = self._hashabledict()
+        else:
+            # Use hashable dictionaries for caching the solve
+            d = self._hashabledict()
+            for key, value in zip(self.jax_inputs.keys(), inputs_values):
+                d[key] = value
+        # Solver
+        logging.debug("_jaxify_solve:")
```

Comment:
Are these meant to be here? I notice you aren't using the normal pybamm logger, which is fair enough since they are quite low level.

Reply:
I can see the utility for creating a dev logger for this solver, but it should be given a name and documented somewhere.

Reply:
I've named the logger and made reference to it in the class docstring.